Deprecate synced flush

dnhatn opened the pull request:
A normal flush has the same effect as a synced flush on Elasticsearch 7.6 or later. It's deprecated in 7.6 and will be removed in 8.0.

Relates #50776

dnhatn committed:
Deprecate synced flush

elasticmachine commented:
Pinging @elastic/es-distributed (:Distributed/Distributed)

jrodewig commented on client/rest-high-level/src/main/java/org/elasticsearch/client/IndicesClient.java:
```diff
@@ -423,7 +423,10 @@ public Cancellable flushAsync(FlushRequest flushRequest, RequestOptions options,
      * @return the response
      * @throws IOException in case there is a problem sending the request or parsing back the response
+     * @deprecated synced flush is deprecated and will be removed in future Elasticsearch versions.
+     * Please use normal {@link #flush(FlushRequest, RequestOptions)} instead.
```
```suggestion
     * Use {@link #flush(FlushRequest, RequestOptions)} instead.
```

jrodewig commented on client/rest-high-level/src/main/java/org/elasticsearch/client/IndicesClient.java:
```diff
@@ -437,7 +440,10 @@ public SyncedFlushResponse flushSynced(SyncedFlushRequest syncedFlushRequest, Re
      * @param listener the listener to be notified upon request completion
      * @return cancellable that may be used to cancel the request
+     * @deprecated synced flush is deprecated and will be removed in future Elasticsearch versions.
+     * Please use normal {@link #flushAsync(FlushRequest, RequestOptions, ActionListener)} instead.
```
```suggestion
     * Use {@link #flushAsync(FlushRequest, RequestOptions, ActionListener)} instead.
```

jrodewig commented on client/rest-high-level/src/main/java/org/elasticsearch/client/IndicesClient.java:
```diff
@@ -437,7 +440,10 @@ public SyncedFlushResponse flushSynced(SyncedFlushRequest syncedFlushRequest, Re
      * @param options the request options (e.g. headers), use {@link RequestOptions#DEFAULT} if nothing needs to be customized
      * @param listener the listener to be notified upon request completion
      * @return cancellable that may be used to cancel the request
+     * @deprecated synced flush is deprecated and will be removed in future Elasticsearch versions.
```
```suggestion
     * @deprecated Synced flush is deprecated and will be removed in a future version.
```

jrodewig commented on client/rest-high-level/src/test/java/org/elasticsearch/client/documentation/IndicesClientDocumentationIT.java:
```diff
@@ -1018,7 +1018,10 @@ public void testSyncedFlushIndex() throws Exception {
             // tag::flush-synced-execute
-            SyncedFlushResponse flushSyncedResponse = client.indices().flushSynced(request, RequestOptions.DEFAULT);
+            SyncedFlushResponse flushSyncedResponse = client.indices().flushSynced(request, expectWarnings(
+                "Synced flush is deprecated and will be removed in future Elasticsearch versions. " +
```
```suggestion
                "Synced flush is deprecated and will be removed in a future version. " +
```

jrodewig commented on client/rest-high-level/src/test/java/org/elasticsearch/client/documentation/IndicesClientDocumentationIT.java:
```diff
@@ -1018,7 +1018,10 @@ public void testSyncedFlushIndex() throws Exception {
-            SyncedFlushResponse flushSyncedResponse = client.indices().flushSynced(request, RequestOptions.DEFAULT);
+            SyncedFlushResponse flushSyncedResponse = client.indices().flushSynced(request, expectWarnings(
+                "Synced flush is deprecated and will be removed in future Elasticsearch versions. " +
+                "Please use normal flush at _/flush or /{index}/_flush instead."
```
```suggestion
                "Use _/flush or /{index}/_flush instead."
```

jrodewig commented on client/rest-high-level/src/test/java/org/elasticsearch/client/documentation/IndicesClientDocumentationIT.java:
```diff
@@ -1062,7 +1065,10 @@ public void onFailure(Exception e) {
-            client.indices().flushSyncedAsync(request, RequestOptions.DEFAULT, listener); // <1>
+            client.indices().flushSyncedAsync(request, expectWarnings(
+                "Synced flush is deprecated and will be removed in future Elasticsearch versions. " +
+                "Please use normal flush at _/flush or /{index}/_flush instead."
```
```suggestion
                "Use _/flush or /{index}/_flush instead."
```

jrodewig commented on client/rest-high-level/src/test/java/org/elasticsearch/client/documentation/IndicesClientDocumentationIT.java:
```diff
@@ -1062,7 +1065,10 @@ public void onFailure(Exception e) {
             // tag::flush-synced-execute-async
-            client.indices().flushSyncedAsync(request, RequestOptions.DEFAULT, listener); // <1>
+            client.indices().flushSyncedAsync(request, expectWarnings(
+                "Synced flush is deprecated and will be removed in future Elasticsearch versions. " +
```
```suggestion
                "Synced flush is deprecated and will be removed in future a version. " +
```

jrodewig commented on docs/reference/indices/synced-flush.asciidoc:
```diff
@@ -12,6 +12,10 @@ POST /twitter/_flush/synced
 --------------------------------------------------
 // TEST[skip: Synced flush can conflict with scheduled flushes in doc tests]
 
+deprecated:[7.6, synced-flush is deprecated and will be removed in future
```
```suggestion
deprecated:[7.6, Synced flush is deprecated and will be removed in a future
```

jrodewig commented on docs/reference/indices/synced-flush.asciidoc:
```diff
@@ -12,6 +12,10 @@ POST /twitter/_flush/synced
 // TEST[skip: Synced flush can conflict with scheduled flushes in doc tests]
 
+deprecated:[7.6, synced-flush is deprecated and will be removed in future
+Elasticsearch versions. Please use normal <<indices-flush,flush>> instead.
```
```suggestion
version. Use <<indices-flush,flush>> instead.
```

jrodewig commented on docs/reference/indices/synced-flush.asciidoc:
```diff
@@ -12,6 +12,10 @@ POST /twitter/_flush/synced
 
+deprecated:[7.6, synced-flush is deprecated and will be removed in future
+Elasticsearch versions. Please use normal <<indices-flush,flush>> instead.
+A normal <<indices-flush,flush>> has the same effect as a synced flush on
```
```suggestion
Flush has the same effect as synced flush in
```

jrodewig commented on docs/reference/indices/synced-flush.asciidoc:
```diff
@@ -12,6 +12,10 @@ POST /twitter/_flush/synced
+deprecated:[7.6, synced-flush is deprecated and will be removed in future
+Elasticsearch versions. Please use normal <<indices-flush,flush>> instead.
+A normal <<indices-flush,flush>> has the same effect as a synced flush on
+Elasticsearch 7.6 or later]
```
```suggestion
{es} 7.6 and later versions.]
```

jrodewig commented on docs/reference/upgrade/synced-flush.asciidoc:
```diff
@@ -3,8 +3,13 @@
 operations are listed in the response body, although the request itself
 still returns a 200 OK status. If there are failures, reissue the request.
+
+Note that synced flush is deprecated and will be removed in future versions.
```
```suggestion
NOTE: Synced flush is deprecated and will be removed in a future version.
```

jrodewig commented on docs/reference/upgrade/synced-flush.asciidoc:
```diff
@@ -3,8 +3,13 @@
 still returns a 200 OK status. If there are failures, reissue the request.
+
+Note that synced flush is deprecated and will be removed in future versions.
+A normal flush has the same effect as a synced flush on Elasticsearch 7.6
```
```suggestion
Flush has the same effect as synced flush in {es} 7.6
```

jrodewig commented on docs/reference/upgrade/synced-flush.asciidoc:
```diff
@@ -3,8 +3,13 @@
+
+Note that synced flush is deprecated and will be removed in future versions.
+A normal flush has the same effect as a synced flush on Elasticsearch 7.6
+or later.
```
```suggestion
and later versions.
```

jrodewig commented on rest-api-spec/src/main/resources/rest-api-spec/api/indices.flush_synced.json:
```diff
@@ -2,7 +2,7 @@
     "documentation":{
       "url":"https://www.elastic.co/guide/en/elasticsearch/reference/master/indices-synced-flush-api.html",
-      "description":"Performs a synced flush operation on one or more indices."
+      "description":"Performs a synced flush operation on one or more indices. Synced flush is deprecated and will be removed in future Elasticsearch versions. Use normal flush instead"
```
```suggestion
      "description":"Performs a synced flush operation on one or more indices. Synced flush is deprecated and will be removed in a future version. Use flush instead."
```

jrodewig commented on rest-api-spec/src/main/resources/rest-api-spec/test/indices.flush/10_basic.yml:
```diff
@@ -1,26 +1,33 @@
+  - do:
+      warnings:
+        - Synced flush is deprecated and will be removed in future Elasticsearch versions.
+          Please use normal flush at _/flush or /{index}/_flush instead.
```
```suggestion
          Please use  _/flush or /{index}/_flush instead.
```

jrodewig commented on rest-api-spec/src/main/resources/rest-api-spec/test/indices.flush/10_basic.yml:
```diff
@@ -1,26 +1,33 @@
+        wait_for_status: green
+  - do:
+      warnings:
+        - Synced flush is deprecated and will be removed in future Elasticsearch versions.
```
```suggestion
        - Synced flush is deprecated and will be removed in a future version.
```

jrodewig commented on server/src/main/java/org/elasticsearch/action/admin/indices/flush/TransportSyncedFlushAction.java:
```diff
@@ -32,7 +34,14 @@
+        new DeprecationLogger(LogManager.getLogger(TransportSyncedFlushAction.class));
+
+    public static final String SYNCED_FLUSH_DEPRECATION_MESSAGE =
+        "Synced flush is deprecated and will be removed in future Elasticsearch versions. " +
```
```suggestion
        "Synced flush is deprecated and will be removed in a future version. " +
```

jrodewig commented on server/src/main/java/org/elasticsearch/action/admin/indices/flush/TransportSyncedFlushAction.java:
```diff
@@ -32,7 +34,14 @@
+
+    public static final String SYNCED_FLUSH_DEPRECATION_MESSAGE =
+        "Synced flush is deprecated and will be removed in future Elasticsearch versions. " +
+            "Please use normal flush at _/flush or /{index}/_flush instead.";
```
```suggestion
            "Please use _/flush or /{index}/_flush instead.";
```

jrodewig reviewed (commented):
Hi @dnhatn!

Overall, I think the docs for this are in pretty good shape. I left several non-blocking nits throughout.

The only blocking change I requested is re: the `deprecated::[]` note in `synced-flush.asciidoc`. The current deprecation block isn't rendering properly. I'd also move it higher up on the page.

dnhatn committed:
Remove in 8.0

dnhatn commented:
Hi @jrodewig Thank you for reviewing. I've addressed your comments.

Discussed with @ywelsch on another channel, we prefer this PR make to 7.x only so we can write the docs more concise.

jrodewig reviewed (approved):
Docs LGTM. Thanks @dnhatn!

ywelsch reviewed (approved):
LGTM

dnhatn commented:
Thanks James and Yannick!
